modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rule/MergeJoinConverterRule.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.query.calcite.rule;
+
+import org.apache.calcite.plan.RelOptCluster;
+import org.apache.calcite.plan.RelOptPlanner;
+import org.apache.calcite.plan.RelOptRule;
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.plan.RelTraitSet;
+import org.apache.calcite.rel.PhysicalNode;
+import org.apache.calcite.rel.RelCollations;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.JoinInfo;
+import org.apache.calcite.rel.logical.LogicalJoin;
+import org.apache.calcite.rel.metadata.RelMetadataQuery;
+import org.apache.ignite.internal.processors.query.calcite.rel.IgniteConvention;
+import org.apache.ignite.internal.processors.query.calcite.rel.IgniteMergeJoin;
+import org.apache.ignite.internal.util.typedef.F;
+
+/**
+ * Ignite Join converter.
+ */
+public class MergeJoinConverterRule extends AbstractIgniteConverterRule<LogicalJoin> {
+    /** */
+    public static final RelOptRule INSTANCE = new MergeJoinConverterRule();
+
+    /**
+     * Creates a converter.
+     */
+    public MergeJoinConverterRule() {
+        super(LogicalJoin.class, "MergeJoinConverterRule");
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean matches(RelOptRuleCall call) {
+        LogicalJoin logicalJoin = call.rel(0);
+
+        return !F.isEmpty(logicalJoin.analyzeCondition().pairs());
```

Comment:
does cortesian and unequality i.e.
_select *
from dealer d1 
join dealer d2  
on d1.id=d2.id AND d1.product>=d2.product_ joins need to be covered here too ?

Reply:
i also don`t know about JoinRelType#SEMI and JoinRelType#ANTI here, also we need to cover all these join types, i hope ...

Reply:
I don't sure we should cover non-equi join with merge algorithm. Merge join is not very optimal in case of many-to-many joins.

Reply:
of course - no, we needn`t cover them, we need to return false in such cases, isn`t it ?

Reply:
Oh, now I see. Yes, you are right. Condition should be like `return !F.isEmpty(logicalJoin.analyzeCondition().pairs()) && logicalJoin.analyzeCondition().isEqui();`

Reply:
fixed